firmware/applications/SignBuddy/imu.c

```diff
@@ -77,45 +92,58 @@ static void bno_init(void)
   ERR_CHECK(bno055_set_operation_mode(BNO055_OPERATION_MODE_NDOF));
 }
 
-static void get_data(void)
+static void imu_task(void *arg)
 {
-  uint32_t ret = 0;
-
-  ret |= bno055_get_sys_calib_stat(&s.sys_calib_stat);
-  ret |= bno055_get_mag_calib_stat(&s.mag_calib_stat);
-  ret |= bno055_get_accel_calib_stat(&s.accel_calib_stat);
-  ret |= bno055_get_gyro_calib_stat(&s.gyro_calib_stat);
-  ret |= bno055_read_quaternion_wxyz(&s.bno055_quat_wxyz);
-  ret |= bno055_read_linear_accel_xyz(&s.bno055_acce_xyz);
-  ERR_CHECK(ret);
+  while (1) {
+    ulTaskNotifyTake(pdFALSE, portMAX_DELAY);
+    sample_data();
+    s.callback();
+  }
+}
 
-  LOG_INFO("Syst calib: %d\r\n", s.sys_calib_stat);
-  LOG_INFO("Magt calib: %d\r\n", s.mag_calib_stat);
-  LOG_INFO("Accl calib: %d\r\n", s.accel_calib_stat);
-  LOG_INFO("Gyro calib: %d\r\n", s.gyro_calib_stat);
-  LOG_INFO("Quat dataW: %d\r\n", s.bno055_quat_wxyz.w);
-  LOG_INFO("Quat dataX: %d\r\n", s.bno055_quat_wxyz.x);
-  LOG_INFO("Quat dataY: %d\r\n", s.bno055_quat_wxyz.y);
-  LOG_INFO("Quat dataZ: %d\r\n", s.bno055_quat_wxyz.z);
-  LOG_INFO("Acce dataX: %d\r\n", s.bno055_acce_xyz.x);
-  LOG_INFO("Acce dataY: %d\r\n", s.bno055_acce_xyz.y);
-  LOG_INFO("Acce dataZ: %d\r\n", s.bno055_acce_xyz.z);
+static void get_data(void)
+{
+  s.imu_values[0] = (int16_t) s.sys_calib_stat;
+  s.imu_values[1] = (int16_t) s.mag_calib_stat;
+  s.imu_values[2] = (int16_t) s.accel_calib_stat;
+  s.imu_values[3] = (int16_t) s.gyro_calib_stat;
+  s.imu_values[4] = s.bno055_euler_hrp.h;
+  s.imu_values[5] = s.bno055_euler_hrp.r;
+  s.imu_values[6] = s.bno055_euler_hrp.p;
+  s.imu_values[7] = s.bno055_acce_xyz.x;
+  s.imu_values[8] = s.bno055_acce_xyz.y;
+  s.imu_values[9] = s.bno055_acce_xyz.z;
```

Comment:
why are we storing these as an array, when they have meaningful names ?

Reply:
i am not sure how to iterate over a struct and some of them are 8 bits and others are 16 bits. What do you suggest?

Reply:
just store them with the meaningful names they have
maybe wrapped in a imu_data struct

Reply:
Yeah, just store them directly in the proto imu data type, and then pass a pointer into get data and copy the data to the pointer